src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala

```diff
@@ -0,0 +1,216 @@
+                 tr_currency:              Option[String],
+                 tr_total_base:            Option[Double],
+                 tr_tax_base:              Option[Double],
+                 tr_shipping_base:         Option[Double],
+                 ti_currency:              Option[String],
+                 ti_price_base:            Option[Double],
+                 base_currency:            Option[String],
+                 geo_timezone:             Option[String],
+                 mkt_clickid:              Option[String],
+                 mkt_network:              Option[String],
+                 etl_tags:                 Option[String],
+                 dvce_sent_tstamp:         Option[Instant],
+                 refr_domain_userid:       Option[String],
+                 refr_device_tstamp:       Option[Instant],
+                 derived_contexts:         Contexts,
+                 domain_sessionid:         Option[String],
+                 derived_tstamp:           Option[Instant],
+                 event_vendor:             Option[String],
+                 event_name:               Option[String],
+                 event_format:             Option[String],
+                 event_version:            Option[String],
+                 event_fingerprint:        Option[String],
+                 true_tstamp:              Option[Instant]) {
+
+  def inventory: Set[Data.ShreddedType] = {
+    val unstructEvent = unstruct_event
+      .data
+      .toSet
+      .map((ue: SelfDescribingData[Json]) => Data.ShreddedType(Data.UnstructEvent, ue.schema))
+
+    val derivedContexts = derived_contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.DerivedContexts), ctx.schema))
+
+    val customContexts = contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.CustomContexts), ctx.schema))
+
+    customContexts ++ derivedContexts ++ unstructEvent
+  }
+
+  def toJson(lossy: Boolean): Json =
+    if (lossy) {
+      // Convert to lossless JSON
+      val json = this.asJsonObject.toMap
+
+      // Filter out auto-derived contexts and transform unstruct_event
+      val transformed =
+        json
+          .flatMap {
+            case ("unstruct_event", _) => transformUnstruct(unstruct_event)
+            case (k, v) => Some((k, v))
+          }
+          .filterNot(_._1 == "contexts")
+          .filterNot(_._1 == "derived_contexts")
+
+      // Build geo_location from latitude/longitude
+      val geoLocation = for {
+        lat <- json.find(_._1 == "geo_latitude")
+        lon <- json.find(_._1 == "geo_longitude")
+      } yield s"${lat._2},${lon._2}"
+
+      // Add "lossy" contexts and geo_location
+      val transformedWithContexts = geoLocation match {
+        case Some(geo) => transformed ++ transformContexts(derived_contexts) ++ transformContexts(contexts) + ("geo_location" -> geo.asJson)
```

Comment:
What's the point of writing `transformed ++ transformContexts(derived_contexts) ++ transformContexts(contexts)` twice?